File: android/core/src/main/kotlin/app/kidschedule/core/reminder/Reminder.kt
package app.kidschedule.core.reminder

import app.kidschedule.core.Protocol

// 提醒阈值与下次触发时刻计算,协议 §13。时间均为 epoch 毫秒。

data class ReminderEvent(
    val startedAt: Long,
    val endedAt: Long? = null,
    val ongoing: Boolean = false,
    val deleted: Boolean = false,
)

enum class ActivityKind { INSTANT, DURATION }
enum class ReminderMode { AUTO, FIXED, OFF }

data class ReminderResult(val thresholdSec: Long?, val nextFireAtMillis: Long?) {
    companion object {
        val NONE = ReminderResult(null, null)
    }
}

object ReminderCalculator {

    fun compute(
        kind: ActivityKind,
        mode: ReminderMode,
        fixedIntervalSec: Long?,
        events: List<ReminderEvent>,
    ): ReminderResult {
        if (mode == ReminderMode.OFF) return ReminderResult.NONE

        val sample = events
            .filter { !it.deleted }
            .sortedBy { it.startedAt }
            .takeLast(Protocol.REMINDER_SAMPLE_N)

        if (sample.isEmpty()) return ReminderResult.NONE
        if (sample.any { it.ongoing }) return ReminderResult.NONE

        val thresholdSec: Long = when (mode) {
            ReminderMode.FIXED -> fixedIntervalSec ?: return ReminderResult.NONE
            ReminderMode.AUTO -> {
                if (sample.size < Protocol.REMINDER_MIN_SAMPLES) return ReminderResult.NONE
                val intervals = sample.zipWithNext { a, b -> (b.startedAt - a.startedAt) / 1000 }.sorted()
                // P90 = 升序第 ceil(0.9n) 个(1-based);整数运算避免浮点误差
                val idx = (9 * intervals.size + 9) / 10
                intervals[idx - 1]
            }
            ReminderMode.OFF -> return ReminderResult.NONE
        }

        val last = sample.last()
        val anchor = when (kind) {
            ActivityKind.INSTANT -> last.startedAt
            ActivityKind.DURATION -> last.endedAt ?: last.startedAt
        }
        return ReminderResult(thresholdSec, anchor + thresholdSec * 1000)
    }
}

File: ios/KidSchedule/Core/Reminder.swift
import Foundation

// 提醒阈值与下次触发时刻计算,协议 §13。时间均为 epoch 毫秒。

struct ReminderEvent {
    let startedAt: Int64
    var endedAt: Int64? = nil
    var ongoing: Bool = false
    var deleted: Bool = false
}

enum ActivityKind: String {
    case instant, duration
}

enum ReminderMode: String {
    case auto, fixed, off
}

struct ReminderResult: Equatable {
    let thresholdSec: Int64?
    let nextFireAtMillis: Int64?

    static let none = ReminderResult(thresholdSec: nil, nextFireAtMillis: nil)
}

enum ReminderCalculator {

    static func compute(
        kind: ActivityKind,
        mode: ReminderMode,
        fixedIntervalSec: Int64?,
        events: [ReminderEvent]
    ) -> ReminderResult {
        if mode == .off { return .none }

        let sample = events
            .filter { !$0.deleted }
            .sorted { $0.startedAt < $1.startedAt }
            .suffix(SyncProtocol.reminderSampleN)

        if sample.isEmpty { return .none }
        if sample.contains(where: { $0.ongoing }) { return .none }

        let thresholdSec: Int64
        switch mode {
        case .fixed:
            guard let fixed = fixedIntervalSec else { return .none }
            thresholdSec = fixed
        case .auto:
            if sample.count < SyncProtocol.reminderMinSamples { return .none }
            var intervals: [Int64] = []
            for (a, b) in zip(sample, sample.dropFirst()) {
                intervals.append((b.startedAt - a.startedAt) / 1000)
            }
            intervals.sort()
            // P90 = 升序第 ceil(0.9n) 个(1-based);整数运算避免浮点误差
            let idx = (9 * intervals.count + 9) / 10
            thresholdSec = intervals[idx - 1]
        case .off:
            return .none
        }

        let last = sample.last!
        let anchor: Int64
        switch kind {
        case .instant: anchor = last.startedAt
        case .duration: anchor = last.endedAt ?? last.startedAt
        }
        return ReminderResult(thresholdSec: thresholdSec, nextFireAtMillis: anchor + thresholdSec * 1000)
    }
}
